### PR TITLE
Add colon escape logic to config add

### PR DESCRIPTION
**Pull request by johnwparent**
`process_config_path` disregards typical yaml escape syntax, and as such, Windows paths containing drives are broken down by the yaml separator `:`, which is also the drive separator. This PR implements a parser that respects yaml quoting escapes and replaces any ':' characters within the quotes with a placeholder character, which is replaced post processing. 

Verified that this works for the CI.

### CODE REVIEW

**Comment by haampie** [comment]
% already has a meaning inside specs, and specs appear in config, so I don't think it's the best escape character. Isn't quoting enough? I think that covers more cases generally, with the slight downside of shell escape issues `spack config add 'abc:"def ghi"'` (but I guess linux people are fine with that given ``ssh <host> <command>``)

**Comment by johnwparent** [comment]
> % already has a meaning inside specs, and specs appear in config, so I don't think it's the best escape character.

Fair point, definitely some surface to accidentally interpret a compiler spec as an escape  (although the compiler would have to be named '58' for that to be the case, which feels unlikely). Perhaps a `bell` character or similar would be better?

>  Isn't quoting enough? I think that covers more cases generally, with the slight downside of shell escape issues `spack config add 'abc:"def ghi"'` (but I guess linux people are fine with that given `ssh <host> <command>`)

Unfortunately, no, quoting is not sufficient because `lib/spack/spack/config.py::process_config_path` does not actually use yaml formatting when parsing, it just naively breaks the string down by colon separators, regardless of where the quotes occur. So something like `spack config add config:install_tree:root:"C:/example/path"` would be parsed as: `{ config: { install_tree: { root: { "C :  /example/path" }}}}` which is incorrect and should be: `{ config: { install_tree: { root: "C:/example/path" }}}`

I could refactor the `process_config_path` method but that's a more involved task and my goal here was to allow the Windows Gitlab CI team to make meaningful forward progress. Now that they have this patch and it works, that's a viable option.

**Comment by scheibelp** [comment]
I'm also inclined to want to use quoting vs. introducing a new escape sigil, the following diff appears to work for me:

```
diff --git a/lib/spack/spack/config.py b/lib/spack/spack/config.py
index b3fb5648ac..6891d4a61f 100644
--- a/lib/spack/spack/config.py
+++ b/lib/spack/spack/config.py
@@ -1263,6 +1263,16 @@ def process_config_path(path):
             front.append = True
 
         result.append(front)
+
+        quote = "['\"]"
+        not_quote = "[^'\"]"
+        if re.match(f"^{quote}", path):
+            m = re.match(rf"^{quote}({not_quote}+){quote}$", path)
+            if not m:
+                raise ValueError("Quotes indicate value, but there are additional path entries")
+            result.append(m.group(1))
+            break
+
     return result
```

with that I can do:

```
>>> from spack.config import process_config_path
>>> process_config_path('hi:there:"C://a/path"')
['hi', 'there', 'C://a/path']
```

does that seem reasonable? Or have I perhaps missed some edge cases that are covered in your tests?

**Comment by johnwparent** [comment]
> I'm also inclined to want to use quoting vs. introducing a new escape sigil, the following diff appears to work for me:
> 
> ```
> diff --git a/lib/spack/spack/config.py b/lib/spack/spack/config.py
> index b3fb5648ac..6891d4a61f 100644
> --- a/lib/spack/spack/config.py
> +++ b/lib/spack/spack/config.py
> @@ -1263,6 +1263,16 @@ def process_config_path(path):
>              front.append = True
>  
>          result.append(front)
> +
> +        quote = "['\"]"
> +        not_quote = "[^'\"]"
> +        if re.match(f"^{quote}", path):
> +            m = re.match(rf"^{quote}({not_quote}+){quote}$", path)
> +            if not m:
> +                raise ValueError("Quotes indicate value, but there are additional path entries")
> +            result.append(m.group(1))
> +            break
> +
>      return result
> ```
> 
> with that I can do:
> 
> ```
> >>> from spack.config import process_config_path
> >>> process_config_path('hi:there:"C://a/path"')
> ['hi', 'there', 'C://a/path']
> ```
> 
> does that seem reasonable? Or have I perhaps missed some edge cases that are covered in your tests?

This seems to work as far as I can tell. Are the config modifications applied to the user/env scoped config file? 

One caveat is that this assumes anything inside of double quotes will be the last thing parsed, and purely based on what is legal in yaml, that's not always true.
If we don't care about that, that's fine, but in we had a key with a space (or some other character that needs escaping for some reason), this approach would preclude that from working.

I have not tested this on the CI but I did test it against my local powershell with a similar env variable. The command succeeded but I want to make sure the command created the correct path, how would I verify that?

**Comment by scheibelp** [comment]
> Are the config modifications applied to the user/env scoped config file?

The scope is chosen as an option to `spack config`, like `spack --scope=... config add...` (this gets passed into `config.add()`).

> One caveat is that this assumes anything inside of double quotes will be the last thing parsed, and purely based on what is legal in yaml, that's not always true.

More than assumes, this enforces that. It is more strict than YAML, but a path is always `a:nested:set:of:keys:and:then:finally:a:value`: given that none of our keys contains ":", I think it's reasonable to enforce that likewise no keys can have `"`.

> I have not tested this on the CI but I did test it against my local powershell with a similar env variable. The command succeeded but I want to make sure the command created the correct path, how would I verify that?

You can use `config.get` (e.g. in a unit test)

Also, a quick read through `spack.config.add()` indicates you'll have to update one other spot:

```
    if has_existing_value:
        path, _, value = fullpath.rpartition(":")
```

we should just grab `components[-1]` instead (and probably update the docstring of `process_config_path`).

**Comment by johnwparent** [comment]
@scheibelp this is ready for review.

File: lib/spack/spack/config.py [hunk 1]
```diff
@@ -889,7 +889,10 @@ def add(fullpath, scope=None):
             break
 
     if has_existing_value:
-        path, _, value = fullpath.rpartition(":")
+        path, value, _ = fullpath.rpartition(components[-1])
+        # Breaking down path in this way leaves a dangling
+        # separator, remove it
+        path, *_ = path.rpartition(":")
```
Comment (scheibelp) [inline lib/spack/spack/config.py]: Can we do the prior two lines without `rpartition`?

```
path = ":".join(components[:-1]
value = components[-1]
```

IMO this is a confusing use of rpartition (usually the middle value is discarded, but in this case we are using it like a group re, which is atypical).

File: lib/spack/spack/config.py [hunk 2]
```diff
@@ -1262,6 +1271,16 @@ def process_config_path(path):
             front = syaml.syaml_str(front)
             front.append = True
 
+        quote = "['\"]"
+        not_quote = "[^'\"]"
```
Comment (scheibelp) [inline lib/spack/spack/config.py:1273]: (no change needed) will this allow `'inconsistentquotes"` or would YAML complain? I don't think it's necessary to enforce if YAML allows (since we get rid of them)

**Review by scheibelp — CHANGES_REQUESTED** [review]
One request and one question, hopefully both minor.

File: lib/spack/spack/config.py (hunk 2 above)
Comment (johnwparent) [reply lib/spack/spack/config.py:1273]: I think we're divorcing from yaml conventions with this operation sufficiently in the first place that it doesn't matter from the user perspective and the quotes will be stripped before any actual yaml parsers see it, so while, yes it would allow for that, IMO it's a non issue.

File: lib/spack/spack/config.py (hunk 1 above)
Comment (johnwparent) [reply lib/spack/spack/config.py]: This approach wont work. For example (from test case `cmd/config.py::test_config_add_override_leaf`):
This approach produces
```
# spack config add "config:template_dirs::[test2]"
path = ":".join(components[-1]) # path = config:template
value = components[-1] # components = [test2]
```
Whereas what's expected is:
```
# spack config add "config:template_dirs::[test2]"
path, value, _ = fullpath.rpartition(components[-1]) # path = 'config:template_dirs::', value = [test2]
path, *_ = path.rpartition(":") # path = 'config:template_dirs:'
```
As you can see, the approach suggested strips off a crucial dangling `:`. 

Alternatively we could write:
```
path = ":".join(fullpath.split(":")[:-1])
value = components[-1]
```

which IMO is far less obvious than the `rpartition`.

Comment (scheibelp) [reply lib/spack/spack/config.py]: I don't think the dangling colon is crucial: Your test for example will result in `path = "config:install_tree:root"` (no last colon). The approach I suggested will replace `::` with `:` in the middle of the path, but that isn't necessary for access (which is what is being done in this context).

Are you saying that when you try `spack config add config:install_tree:root::"example"` that it fails unless the path is `config:install_tree:root:`?

Whatever happens, it doesn't make sense to set `value` using rpartition, it is just `components[-1]`. I'm also surprised it works to `rpartition` with `components[-1]` at all since that omits the quotes (e.g. you'd be left with `example:config:path:"` as the string). I'm inclined to say that it would be preferable to avoid it for `path` as well, but if I'm overlooking some case where an intermediate `::` is required for access, I can see it having more appeal. In that case though, I'd rather `process_config_path` got updated to return this preserved string (that function should handle breaking things up into sensible parts and users of that function shouldn't have to worry about reassembling them).

Comment (johnwparent) [reply lib/spack/spack/config.py]: The dangling `:` is necessary for overrides. The `install_tree` example doesn't use overrides, but the test I referenced does. If I remove the dangling colon, all override related tests that pass through the `add` interface fail because we're not processing them as overrides due to the dropped `:`. 
In the docstring of `process_config_path` : 
`Process a path argument to config.set() that may contain overrides ('::' or trailing ':')`
Where the overrides are denoted by the trailing `:`. It is an expected and established part of that interface. 

Sure, we don't need to use `rpartition` to set value, I was more just taking advantage of a convenient side effect there, very much not a crucial piece of code. 
We can just go with my alternative suggestion if we want to avoid `rpartition` altogether. 

I disagree that this should ever be the responsibility of `process_config_path`. As you said, that method is responsible for breaking down the "paths" into sensible component parts, and as such, should not be responsible for reconstructing them re: single responsibility. If the method is responsible for breaking things down, the caller should reasonably expect to have to reconstruct things if they need things reconstructed. 
Additionally, the caller already has the original string, they called the method with it, and reconstructing based on some arbitrary heuristic in `process_config_path` doesn't really seem globally useful to callers as opposed to allowing them reconstruct their own strings based on contextual need. 
My example also doesn't reconstruct the return of `process_config_path` it uses the original string to understand the path outside the context of the value, which `process_config_path` successfully decomposes into something clear and accessible. 


Edit: this is not scoped only to dangling colons, but also intermediary double colons like `config::template_dirs:[test2]` that we rely on would be blown away by the suggested approach to reconstruction.

File: lib/spack/spack/config.py (hunk 2 above)
Comment (johnwparent) [reply lib/spack/spack/config.py:1273]: If you're worried about it, I can just add a a group reference to the regex, no problem.

Comment (scheibelp) [reply lib/spack/spack/config.py:1273]: I'll mark this as resolved: it doesn't need to be altered.

Comment (johnwparent) [reply lib/spack/spack/config.py:1273]: Sounds good to me. Thanks!

File: lib/spack/spack/config.py (hunk 1 above)
Comment (scheibelp) [reply lib/spack/spack/config.py]: The following diff applied to the tip of this PR avoids the use of rpartition:

```
diff --git a/lib/spack/spack/config.py b/lib/spack/spack/config.py
index 79e33fc21a..b0eb52d019 100644
--- a/lib/spack/spack/config.py
+++ b/lib/spack/spack/config.py
@@ -888,12 +888,11 @@ def add(fullpath, scope=None):
                 value = {component: value}
             break
 
+    if override:
+        path += "::"
+
     if has_existing_value:
-        path, value, _ = fullpath.rpartition(components[-1])
-        # Breaking down path in this way leaves a dangling
-        # separator, remove it
-        path, *_ = path.rpartition(":")
-        value = syaml.load_config(value)
+        value = syaml.load_config(components[-1])
         existing = get(path, scope=scope)
```

and appears to pass the config tests. I noticed there is already a reassembly of `path` above. I think in fact 

`value = syaml.load_config(components[-1])`

could be moved above the loop, which might further clarify things. There's a couple instances of variable shadowing which make this harder to reason about.

Comment (johnwparent) [reply lib/spack/spack/config.py]: That does appear to ensure all tests pass, I'll apply it.

Out of curiosity, what is the issue with:
```
path = ":".join(fullpath.split(":")[:-1])
value = components[-1]
```
Which sidesteps the need for logic in the first place?

Edit: properly format the codeblock

Comment (scheibelp) [reply lib/spack/spack/config.py]: > Out of curiosity, what is the issue with:

```
path = ":".join(fullpath.split(":")[:-1])
value = components[-1]
```

I thought at the end of https://github.com/spack/spack/pull/39279#discussion_r1310378871 you were saying that was less obvious than rpartition so:

* I thought rpartition was too complex
* You offered that solution, which to you was also more complex
* I offered another possible solution, hoping it was less complex than the 2 prior options

Perhaps you think this suggestion is also strictly worse? I'm open to other possibilities.

Comment (johnwparent) [reply lib/spack/spack/config.py]: Oh, no, I certainly don't have a better option, it was just a complete lack of using the English language effectively on my part. My apologies.
I had meant to express that sentiment as: "this is harder to read than the `rpartition` option in my opinion, but if we're dropping that, then, this should be our solution". On re-read I... did not communicate that _whatsoever_. My apologies again, I did not mean to push you to spend time writing another alternative. 
I think the solutions are more or less equivalent (although I agree the value assignment should move elsewhere toward the top of the loop regardless, that way we would also be able to drop line 886 in config.py in this PR), although I personally am not a fan of logic statements if we can avoid them.
Either way, I'll leave it  to your discretion. Let me know which you prefer and I'll push (I have git stashes with all the discussed options so its trivial on my end)

Comment (scheibelp) [reply lib/spack/spack/config.py]: In this case, can we do https://github.com/spack/spack/pull/39279#discussion_r1310733035

and 

> I agree the value assignment should move elsewhere toward the top of the loop regardless

since we agree on that, let's move that initial assignment of `value` up

Comment (johnwparent) [reply lib/spack/spack/config.py]: Actually, on a second inspection, due to the nature of the loop, which is setting value to be a potentially nested dictionary, I think we can't set value to `value = syaml.load_config(components[-1])` as an alterative to `value = syaml.load_config(value)` as value may not be equal to `components[-1]` in some cases, unless I'm mistaken?

Comment (johnwparent) [reply lib/spack/spack/config.py]: I've made the other suggested change and pushed.

Comment (scheibelp) [reply lib/spack/spack/config.py]: > Actually, on a second inspection, due to the nature of the loop, which is setting value to be a potentially nested dictionary, I think we can't set value to value = syaml.load_config(components[-1])

That loop starts by setting `value = syaml.load_config(components[-1])`, and we need to set it outside of the loop as well, in 

```
if has_existing_value:
    value = syaml.load_config(value) # <---- right now, value is unset at this point, which I think is triggering the test failures
```

File: lib/spack/spack/config.py [hunk 3]
```diff
@@ -883,13 +884,14 @@ def add(fullpath, scope=None):
...
             for component in reversed(components[idx + 1 : -1]):
                 value = {component: value}
             break
 
+    if override:
+        path += "::"
+
     if has_existing_value:
-        path, _, value = fullpath.rpartition(":")
         value = syaml.load_config(value)
```
Comment (scheibelp) [inline lib/spack/spack/config.py]: This second `value =` should not be required: the one you added at line 866 should be enough. I'm not clear if that is what's causing the test failures though.

**Comment by scheibelp** [comment]
@spackbot run pipeline

**Review by scheibelp — APPROVED** [review]
